Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,470 @@
+  where siz = sizeOf (undefined :: a)
+
+-- | Copy a slice of a mutable primitive array into another, potentially
+-- overlapping array.
+-- The offset and length are given in elements of type @a@.
+movePrimArray :: forall m a. (PrimMonad m, Prim a)
+              => MutablePrimArray (PrimState m) a -- ^ destination array
+              -> Int                              -- ^ offset into destination array
+              -> MutablePrimArray (PrimState m) a -- ^ source array
+              -> Int                              -- ^ offset into source array
+              -> Int                              -- ^ number of prims to copy
+              -> m ()
+{-# INLINE movePrimArray #-}
+movePrimArray (MutablePrimArray dst) doff (MutablePrimArray src) soff n =
+    moveByteArray dst (doff*siz) src (soff*siz) (n*siz)
+  where siz = sizeOf (undefined :: a)
+
+-- | Fill a slice of a mutable primitive array with a value.
+-- The offset and length are given in elements of type @a@.
+setPrimArray :: forall m a. (PrimMonad m, Prim a)
+             => MutablePrimArray (PrimState m) a -- ^ array to fill
+             -> Int                              -- ^ offset into array
+             -> Int                              -- ^ number of values to fill
+             -> a                                -- ^ value to fill with
+             -> m ()
+{-# INLINE setPrimArray #-}
+setPrimArray (MutablePrimArray mba) = setByteArray mba
+
+-- | Resize a primitive array using 'resizeMutableByteArray#'.
+--
+-- To avoid undefined behaviour, the original 'MutablePrimArray' shall not be accessed anymore.
+--
+resizeMutablePrimArray :: forall m a. (PrimMonad m, Prim a)
+                       => MutablePrimArray (PrimState m) a
+                       -> Int
+                       -> m (MutablePrimArray (PrimState m) a)
+resizeMutablePrimArray (MutablePrimArray mba) sz =
+    MutablePrimArray `liftM` resizeMutableByteArray mba (sz * siz)
+  where siz = sizeOf (undefined :: a)
+{-# INLINE resizeMutablePrimArray #-}
+
+-- | Shrink a primitive array using 'shrinkMutableByteArray#'.
+--
+-- The new size argument must be less than or equal to the current size, but it's not checked.
+--
+shrinkMutablePrimArray :: forall m a. (Prim a, PrimMonad m)
+                       => MutablePrimArray (PrimState m) a
+                       -> Int
+                       -> m ()
+shrinkMutablePrimArray (MutablePrimArray mba) sz =
+    shrinkMutableByteArray mba (sz * siz)
+  where siz = sizeOf (undefined :: a)
+{-# INLINE shrinkMutablePrimArray #-}
+
+-- | Check if the two immutable arrays refer to the same memory block.
+samePrimArray :: PrimArray a -> PrimArray a -> Bool
+{-# INLINE samePrimArray #-}
+samePrimArray (PrimArray (ByteArray ba1#)) (PrimArray (ByteArray ba2#)) =
+    isTrue# (sameMutableByteArray# (unsafeCoerce# ba1#) (unsafeCoerce# ba2#))
+
+--------------------------------------------------------------------------------
+
```

Comment:
lest I forget, we can actually construct analogous prefetch operations for boxed arrays (lifted and unlifted flavors both), using the availaable prefetch operations

Reply:
i'll revisit this after i do a proper code review